### ISSUE TITLE
Documentation for right-click options and menu

### ISSUE THREAD

**Issue by tamielbr** [Johni0702/mumble-web#77]
When right-clicking a room there are many options that pop-up. It seems some are activated through ACL (same issue as #74) but other like "copy mumble-web url" are unclear. Can these be explained somewhere to help understand what is made available under what circumstance? I assume that the record button (greyed out) on top is also not yet implemented. Any tips would be useful.

**Comment by Johni0702** [comment]
Pretty much all of the context menu options (including those additionally guarded by ACL) have yet to be implemented (same with the record button).
AFAICT the only context menu options currently implemented are:
- Join Channel (though this is always active, even when you lack permission)
- View/Change/Reset Avatar
- Self Mute/Deafen

**Comment by tamielbr** [comment]
Thanks for the information!

**Comment by Fab1en** [comment]
@Johni0702 I would like to work on the implementation of some missing features. The one I would like to start with is "create new channel". I had a look into the code, and I understand some of it. What I see is that I must first implement the `createChannel` method in [mumble-client](https://github.com/johni0702/mumble-client), is that correct ?

But what is unclear for me is whether it is necessary to implement this functionality on the server-side as well. Is the create Channel already available on the server (at least the heavy client can do it, so why not the web client), or do we need some more websocket-to-64738-tcp-port stuff here ?

**Comment by Johni0702** [comment]
> The one I would like to start with is "create new channel". I had a look into the code, and I understand some of it. What I see is that I must first implement the createChannel method in mumble-client, is that correct ?

That's correct. Your implementation would have to ask the server to create a new channel by sending it a `ChannelState` message with at least `parent` and `name` but without a `channel_id`.
For sending it a ChannelState message, see e.g. https://github.com/Johni0702/mumble-client/blob/22f5b27a978bc878d3ec4b2c72619a8884442b4a/src/channel.js#L69-L75
For info on all the ChannelState fields, see https://github.com/Johni0702/mumble-streams/blob/8b40efe161493a7baace56f979cd114d38d43d66/lib/Mumble.proto#L117-L147

> But what is unclear for me is whether it is necessary to implement this functionality on the server-side as well. Is the create Channel already available on the server (at least the heavy client can do it, so why not the web client), or do we need some more websocket-to-64738-tcp-port stuff here ?

Anything which the normal/heavy client can do over tcp can be done by the web client as well.
At least for the non-webrtc version of the web client, the websocket-to-tcp proxy (websockify) isn't actually aware of the underlying protocol, it literally just wraps the whole stream into a websocket stream. In fact, websockify was written before mumble-web even was a thing, and was originally (afaik) used to build a VNC client in the browser.

**Comment by Fab1en** [comment]
Great, thanks a lot for the explanations.

**Comment by Fab1en** [comment]
I was able to send a ChannelState message, it does not work : I get a permission denied answer. I suspect that this is due to the fact that connecting through the web interface is done whihout authentication, and that only authenticated users can create channels.

Here are the traces : 
request : 
![image](https://user-images.githubusercontent.com/1143066/79101089-1fb13780-7d68-11ea-91f8-1e8a36822b65.png)
answer : 
![image](https://user-images.githubusercontent.com/1143066/79101042-09a37700-7d68-11ea-882d-c5adf6706480.png)

console.log on received message : 
![image](https://user-images.githubusercontent.com/1143066/79101149-3d7e9c80-7d68-11ea-9ac0-178a7a1c1954.png)

Do you know how I can log-in as an authenticated user via the web interface ? Are tokens or password fields useful for that ?

**Comment by Fab1en** [comment]
OK, I found what permission denied type 7 is 
```
// Operation requires certificate.
MissingCertificate = 7;
```
So the question is : is it possible to connect with a certificate via the web interface ?

**Comment by Johni0702** [comment]
Source of the permission denied message is: https://github.com/mumble-voip/mumble/blob/7ed7d95f4bee4a8aac437862cf6d4455de6ca24a/src/murmur/Messages.cpp#L1036-L1039
I.e. looks like you must be either an authenticated user (i.e. have a user id) or have a certificate (hash).

> is it possible to connect with a certificate via the web interface ?

No, the browser does not have control over the certificate used, only websocket to tcp (or rather to tls) proxy does.

> Do you know how I can log-in as an authenticated user via the web interface ? Are tokens or password fields useful for that ?

AFAIK you can't with just vanilla murmur and no ability to set your certificate.
You may be able to log in as an authenticated user when using an [external authenticator](is it possible to connect with a certificate via the web interface ?) though (in that case the user would have to enter a password each time or you could just assign them a new user id (not to be confused with session id) on each connect.

**Comment by Fab1en** [comment]
So, if I understand well, websockify does not use TLS to connect to the mumble server ? It uses its certificate only to make the connection with the browser, is that correct ? So to mumble server point of view, there is no certificate.

In this case, is it possible to make websockify to connect to mumble using TLS ?

Another question : it's not possible to handle the private/public key pair on the browser side, generate a certificate directly in javascript, and ask the mumble server to use it ?

**Comment by Johni0702** [comment]
> websockify does not use TLS to connect to the mumble server

no, it does use TLS but it doesn't use a client certificate (much like the browser accessing any random https page). TLS does not require you to use a client certificate. The normal Mumble client does, browsers don't most of the time and neither does websockify.

> In this case, is it possible to make websockify to connect to mumble using [a client certificate] ?

It might be. Not really sure. That wouldn't be particularly helpful though since then all web clients would be using the same client certificate, i.e. would have the same user id (and only one of them could be connected at any one time).

> Another question : it's not possible to handle the private/public key pair on the browser side, generate a certificate directly in javascript, and ask the mumble server to use it ?

Not really, not without modifying the mumble protocol and server anyway.
Let me explain. This is more or less the steps a mumble packet goes through when being sent from mumble-web to a mumble server:
```
Mumble packet
    Encoded to bytes via protobuf
    | (leaving the JS world, now in the browser)
    |   Wrap in WebSocket packets
    |   |   Wrap in TLS
    |   |   | (leaving the browser)
    |   |   |   TCP packets going over the internet
    |   |   | (entering websockify)
    |   |   Unwrap TLS
    |   Unwrap WebSocket packets
    |   Re-wrap in TLS
    |   | (leaving websockify)
    |   |   TCP packets going over the internet (or maybe also just locally)
    |   | (arriving in the mumble server)
    |   Unwrap TLS (and store client-certificate)
    Decode from bytes via protobuf
Mumble packet
```
The notable thing here is that there are two different TLS connections (one between browser and websockify and one between websockify and the mumble server) and even if we could get the browser to use a client-certificate (which doesn't appear to be possible in a user-friendly way, see second to last paragraph of https://github.com/mumble-voip/grumble/pull/18#issue-163248383), it can never reach the server because websockify needs to unwrap that TLS connection to remove the additional websockify wrapping. So by design of TLS, it cannot forward the browser's client-certificate to the mumble server.

I can only see three ways around that:
- Making the server directly support websockify, thereby removing the websockify-in-the-middle and allowing the server to directly see the browser's client-certificate (not user-friendly though, see above link), this would be nice to have in the long term to make the whole thing easier to set up but it doesn't really solve our problem if user-friendliness is a requirement (which I'd assume it is if we're building a web client)
- Adding some way to do authentication on the mumble protocol layer so it's unaffected by websockify (requires designing or at least finding an existing, secure authentication scheme and implementing it in the mumble-server), is is probably a long-term goal
- Using the existing support for password-based authentication (not to be confused with access tokens) which currently exists in the mumble protocol (the official mumble server only supports password-based authentication with an external authenticator though), this seems somewhat feasible in the short term, hence why I suggested it in my previous comment

**Comment by Fab1en** [comment]
OK, thanks for all your details and the clear schema.

I thought that channel creation would be easy to do, but its not. I will see which external authenticator is best suited for my needs, but that pushes the feature a bit far away I think.

Meanwhile, I will try to concentrate on other features. By the way, I found 2 smalls bugs during this research so I will prepare some tickets about that.

**Comment by Johni0702** [comment]
Oh for completeness sake, there's actually a forth option: Doing TLS in JS as well. This may be difficult to implement (would basically have to compile openssl or something similar to JS) and possibly too slow to execute to be feasible (since we then also need to encrypt all the voice in JS).
That would change the above depiction to look like this:
```
Mumble packet
    Encoded to bytes via protobuf
    |   Wrap in TLS (using generated or manually imported client-certificate)
    |   | (leaving the JS world, now in the browser)
    |   |   Wrap in WebSocket packets
    |   |   |   Wrap in TLS
    |   |   |   | (leaving the browser)
    |   |   |   |   TCP packets going over the internet
    |   |   |   | (entering websockify)
    |   |   |   Unwrap TLS
    |   |   Unwrap WebSocket packets
    |   | (leaving websockify)
    |   |   TCP packets going over the internet (or maybe also just locally)
    |   | (arriving in the mumble server)
    |   Unwrap TLS (and store client-certificate)
    Decode from bytes via protobuf
Mumble packet
```

**Comment by Raddinox** [comment]
Hi,

I just want to say that I have been able to create and delete channels through your mumble-web interface. I did some hacking in your code to enable the menu options and to be able to send a ChannelState message (and ChannelRemove).

But it only works if I log in as a registered user, I used mumble-django to create a user with password and then I log in with my user and password (not the server password). So this works without user-client certificate.

I don't use github as my repo (just created my login to write this) so my repo is at my self-hosted https://git.raddinox.com/forks/mumble-web.git

**Comment by aanze** [comment]
As of 2021, has any implementation of context menu been added ? 
I still can't mute another person as most of the context menu options are greyed.
